File: app/src/main/java/com/bussiness/curemegptapp/util/LinkEncryptionHelper.kt
package com.bussiness.curemegptapp.util

import android.net.Uri
import android.util.Base64

object LinkEncryptionHelper {
    private const val CIPHER_KEY: Int = 0x5A

    /** Encrypts an integer ID into a URL-safe encrypted string token */
    fun encrypt(id: Int): String {
        val stringVal = id.toString()
        val bytes = stringVal.toByteArray(Charsets.UTF_8)
        val encryptedBytes = bytes.map { (it.toInt() xor CIPHER_KEY).toByte() }.toByteArray()
        return Base64.encodeToString(encryptedBytes, Base64.NO_WRAP)
            .replace("+", "-")
            .replace("/", "_")
            .replace("=", "")
    }

    /** Decrypts a URL-safe encrypted token back into an integer ID */
    fun decrypt(string: String): Int? {
        val trimmed = string.trim()
        if (trimmed.isEmpty()) return null

        var base64 = trimmed
            .replace("-", "+")
            .replace("_", "/")

        val remainder = base64.length % 4
        if (remainder > 0) {
            base64 += "=".repeat(4 - remainder)
        }

        return try {
            val data = Base64.decode(base64, Base64.NO_WRAP)
            val decryptedBytes = data.map { (it.toInt() xor CIPHER_KEY).toByte() }.toByteArray()
            val decryptedString = String(decryptedBytes, Charsets.UTF_8)
            decryptedString.toIntOrNull() ?: trimmed.toIntOrNull()
        } catch (e: Exception) {
            trimmed.toIntOrNull()
        }
    }

    /** Encrypts a string into a URL-safe encrypted string token */
    fun encryptString(input: String): String {
        val bytes = input.toByteArray(Charsets.UTF_8)
        val encryptedBytes = bytes.map { (it.toInt() xor CIPHER_KEY).toByte() }.toByteArray()
        return Base64.encodeToString(encryptedBytes, Base64.NO_WRAP)
            .replace("+", "-")
            .replace("/", "_")
            .replace("=", "")
    }

    /** Decrypts a URL-safe encrypted token back into a string */
    fun decryptString(string: String): String? {
        val trimmed = string.trim()
        if (trimmed.isEmpty()) return null

        var base64 = trimmed
            .replace("-", "+")
            .replace("_", "/")

        val remainder = base64.length % 4
        if (remainder > 0) {
            base64 += "=".repeat(4 - remainder)
        }

        return try {
            val data = Base64.decode(base64, Base64.NO_WRAP)
            val decryptedBytes = data.map { (it.toInt() xor CIPHER_KEY).toByte() }.toByteArray()
            val decryptedString = String(decryptedBytes, Charsets.UTF_8)
            decryptedString.ifEmpty { null }
        } catch (e: Exception) {
            Uri.decode(string)
        }
    }
}

File: temp_CureMeGPTRAvi/CureMeGPTRAvi 6/CureMeGPT/HelperFile/LinkEncryptionHelper.swift
//
//  LinkEncryptionHelper.swift
//  CureMeGPT
//
//  Created by YES IT Labs on 27/07/26.
//

import Foundation

/// Utility helper to encrypt and decrypt deep link parameters (IDs)
struct LinkEncryptionHelper {
    private static let cipherKey: UInt8 = 0x5A
    
    /// Encrypts an integer ID into a URL-safe encrypted string token
    static func encrypt(id: Int) -> String {
        let stringVal = "\(id)"
        guard let data = stringVal.data(using: .utf8) else { return "\(id)" }
        let encryptedBytes = data.map { $0 ^ cipherKey }
        let base64 = Data(encryptedBytes).base64EncodedString()
        return base64
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
    
    /// Decrypts a URL-safe encrypted token back into an integer ID
    /// Tries decryption first, with fallback to direct integer parsing for unencrypted legacy links
    static func decrypt(string: String) -> Int? {
        let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return nil }
        
        // 1. Base64 URL-safe decoding & XOR decryption
        var base64 = trimmed
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        
        let remainder = base64.count % 4
        if remainder > 0 {
            base64.append(String(repeating: "=", count: 4 - remainder))
        }
        
        if let data = Data(base64Encoded: base64) {
            let decryptedBytes = data.map { $0 ^ cipherKey }
            if let decryptedString = String(bytes: decryptedBytes, encoding: .utf8),
               let decryptedInt = Int(decryptedString) {
                return decryptedInt
            }
        }
        
        // 2. Direct integer fallback for unencrypted legacy links
        return Int(trimmed)
    }
    
    /// Encrypts a string into a URL-safe encrypted string token
    static func encryptString(_ string: String) -> String {
        guard let data = string.data(using: .utf8) else { return string }
        let encryptedBytes = data.map { $0 ^ cipherKey }
        let base64 = Data(encryptedBytes).base64EncodedString()
        return base64
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
    
    /// Decrypts a URL-safe encrypted token back into a string
    static func decryptString(_ string: String) -> String? {
        let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return nil }
        
        var base64 = trimmed
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        
        let remainder = base64.count % 4
        if remainder > 0 {
            base64.append(String(repeating: "=", count: 4 - remainder))
        }
        
        if let data = Data(base64Encoded: base64) {
            let decryptedBytes = data.map { $0 ^ cipherKey }
            if let decryptedString = String(bytes: decryptedBytes, encoding: .utf8), !decryptedString.isEmpty {
                return decryptedString
            }
        }
        
        return string.removingPercentEncoding ?? string
    }
}
